diff -ur dnspython-1.10.0/dns/name.py dnspython-1.10.0-new/dns/name.py
--- dnspython-1.10.0/dns/name.py	2011-07-09 10:05:21.000000000 -0400
+++ dnspython-1.10.0-new/dns/name.py	2014-08-26 09:28:08.000000000 -0400
@@ -419,7 +419,7 @@
             else:
                 if not compress is None and len(n) > 1:
                     pos = file.tell()
-                    if pos < 0xc000:
+                    if pos <= 0x3fff:
                         compress[n] = pos
                 l = len(label)
                 file.write(chr(l))
